File: sexpr.py
#!/usr/bin/env python
"""
This code is intentionally very c-style
"""

import sys

node_table = []
DEFAULT = "(& (>> t 6) (& (* 2 t) (>> t 1)))"

# node = dict(type="", cval="", ival=0, lval=0, rval=0)

def find_split(s, start, end):
    """
    this function searches for the split between arguments in a binary sexpr
    """
    depth = 0
    for i in range(start, end):
        if depth == 0 and s[i] == ' ':
            return i
        if s[i] == '(':
            depth += 1
        if s[i] == ')':
            depth -= 1
        if depth < 0:
            raise Exception("parse: unmatched ')'")
    if depth > 0:
        raise Exception("parse: unmatched '('")
    return -1

def parse(s, start, end):
    """
    we are going to get either an sexpr (wrapped in parens) or an atom
    """
    # number
    if s[start] in "0123456789":
        return dict(type='n', ival=int(s[start:end+1]))

    # variable
    if s[start] == 't' and end-start == 0:
        return dict(type='v')

    # must be an s-exp
    if not (s[start] == '(' and s[end] == ')') or end-start < 4:
        raise Exception("parse: unparsable: '%s'" % str(s[start:end+1]))

    # unitary operator
    if s[start+1] == '~':
        rval = parse(s, start+3, end-1)
        return dict(type='u', cval='~', rval=rval)

    # various binary operators
    if s[start+1] in "*/%+-&^|<>":
        cval = s[start+1]
        offset = 3
        if cval == '<' and s[start+2] == '<' and s[start+3] == ' ':
            cval = 'l'
            offset = 4
        elif cval == '>' and s[start+2] == '>' and s[start+3] == ' ':
            cval = 'r'
            offset = 4
        elif s[start+2] != ' ':
            raise Exception("parse: invalid operator?")
        split = find_split(s, start+offset, end-1)
        lval = parse(s, start+offset, split-1)
        rval = parse(s, split+1, end-1)
        return dict(type='b', cval=cval, lval=lval, rval=rval)

    # should not get here 
    raise Exception("parse: feel through")

def ezparse(s):
    s = s.strip()
    return parse(s, 0, len(s)-1)

def strmachine(sexpr):
    #print sexpr
    # atom
    if sexpr['type'] == 'v':
        return 't'
    elif sexpr['type'] == 'n':
        return str(sexpr['ival'])
    # unary
    elif sexpr['type'] == 'u':
        if sexpr['cval'] == '~':
            return "(~ %s)" % strmachine(sexpr['rval'])
        raise Exception("unexpected unary: %s" % sexpr['cval'])
    # binary
    elif sexpr['type'] == 'b':
        if sexpr['cval'] == 'l' or sexpr['cval'] == 'r':
            oper = '<'
            if sexpr['cval'] == 'r':
                oper = '>'
            return "(%s%s %s %s)" % (oper, oper, strmachine(sexpr['lval']), strmachine(sexpr['rval']))
        elif sexpr['cval'] in '+-*/^|&':
            return "(%s %s %s)" % (sexpr['cval'], strmachine(sexpr['lval']), strmachine(sexpr['rval']))
        raise Exception("unexpected binary: %s" % sexpr['cval'])

def execute(sexpr, t):
    # atom
    if sexpr['type'] == 'v':
        return t
    elif sexpr['type'] == 'n':
        return sexpr['ival']
    # unary
    elif sexpr['type'] == 'u':
        if sexpr['cval'] == '~':
            return ~ execute(sexpr['rval'], t)
        raise Exception("unexpected unary: %s" % sexpr['cval'])
    # binary
    elif sexpr['type'] == 'b':
        if sexpr['cval'] == '+':
            return execute(sexpr['lval'], t) + execute(sexpr['rval'], t)
        if sexpr['cval'] == '-':
            return execute(sexpr['lval'], t) - execute(sexpr['rval'], t)
        if sexpr['cval'] == '*':
            return execute(sexpr['lval'], t) * execute(sexpr['rval'], t)
        if sexpr['cval'] == '/':
            return execute(sexpr['lval'], t) / execute(sexpr['rval'], t)
        if sexpr['cval'] == '^':
            return execute(sexpr['lval'], t) ^ execute(sexpr['rval'], t)
        if sexpr['cval'] == '&':
            return execute(sexpr['lval'], t) & execute(sexpr['rval'], t)
        if sexpr['cval'] == '|':
            return execute(sexpr['lval'], t) | execute(sexpr['rval'], t)
        if sexpr['cval'] == 'r':
            return execute(sexpr['lval'], t) >> execute(sexpr['rval'], t)
        if sexpr['cval'] == 'l':
            return execute(sexpr['lval'], t) << execute(sexpr['rval'], t)
        raise Exception("unexpected binary: %s" % sexpr['cval'])
    pass

def play(machine):
    t = 0
    while True:
        t += 1
        sys.stdout.write(chr(execute(machine, t) & 0x000000FF))

def test_parse():
    tlist = [
        '1',
        't',
        '(+ 1 t)',
        '(~ (~ 1))',
        '(+ (^ t 123) (| 300 t))',
        DEFAULT,
    ]
    for t in tlist:
        assert strmachine(ezparse(t)) == t, "\n%s <- GOT\n%s <- EXPECTED" % (strmachine(ezparse(t)), t)
    flist = [
        '',
        '~',
        '~~1',
        '(+ + +)',
        '(+ 1 1))',
        '(+ 1 1',
    ]
    for f in flist:
        try:
            assert strmachine(ezparse(t)) == "SHOULDFAIL", "%s <- SHOULD HAVE FAILED"% t
        except:
            pass
    print "passed all parse tests!"

def test_execute():
    assert execute(ezparse('t'), 12345) == 12345
    assert execute(ezparse('(+ t 6)'), 5) == 11
    assert execute(ezparse('(& t 99)'), 0x12345) == 0x12345 & 99
    print "passed all execute tests!"

def main():
    if len(sys.argv) <= 1:
        play(ezparse(DEFAULT))
    elif sys.argv[1] == '--test':
        test_parse()
        test_execute()
    elif sys.argv[1] == '--parse':
        if len(sys.argv) < 3:
            print strmachine(ezparse(DEFAULT))
        else:
            print strmachine(ezparse(sys.argv[2]))
    else:
        play(sys.argv[1])

if __name__=='__main__':
    main()

File: sexpr.c

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define DEFAULT "(& (>> t 6) (& (* 2 t) (>> t 1)))"
# define NUMNODE 160

struct node {
    char type;
    char cval;
    unsigned int ival;
    struct node *lval;
    struct node *rval;
};

void print_sexpr(struct node *sexpr);
unsigned int execute(struct node *sexpr, unsigned int t);
void play(struct node *sexpr);
int find_split(char* s, int start, int end);
struct node *parse(char *s, int start, int end);
struct node *new_node(char type, char cval, unsigned int ival,
                      struct node *lval, struct node *rval);
void test();

static struct node node_table[NUMNODE];
static int newest_node = 0;

char inbuffer[256];

int sstrlen(char *s, int max);

int sstrlen(char *s, int max) {
    int i;
    for (i=0; i<max; i++) {
        if (s[i] == '\n' || s[i] == '\0') {
            return i+1;
        }
    }
    return max;
}

void main() {
    int len;
    //test();
    if (stderr == stdout) {
        printf("You are expected to send stderr and stdout different ways.\n");
        exit(-1);
    }
    while (1) {
        fprintf(stderr, "Input a tune in exact s-expr syntax:\n");
        fprintf(stderr, "> ");
        fgets(inbuffer, 256, stdin);    
        len = sstrlen(inbuffer, 256);
        if (len == 256 || len < 1)
            fprintf(stderr, "Invalid input!\n");
        else
            break;
    }
    play(parse(inbuffer, 0, len-2));
    //play(parse(DEFAULT, 0, strlen(DEFAULT)-1));
};

void test() {
    printf(DEFAULT);
    printf("\n");
    print_sexpr(parse(DEFAULT, 0, strlen(DEFAULT)-1));
    printf("\n");
}

int digtoi(char *s, int start, int end) {
    int num = 0;
    for(;start<=end;start++) {
        if (! isdigit(s[start])) {
            fprintf(stderr, "parse: not a digit: %c", s[start]);
            exit(-1);
        }
        num = num*10 + (s[start] - '0');
    }
    return num;
}

struct node *parse(char *s, int start, int end) {
    struct node *lval, *rval;
    char cval;
    int offset, split;

    // number
    if (isdigit(s[start])) {
        return new_node('n', '_', digtoi(s, start, end), NULL, NULL);
    }
    // variable
    if ((s[start] == 't') && (end-start == 0)) {
        return new_node('v', '_', 0, NULL, NULL);
    }

    // must be an s-exp
    if ( !(s[start] == '(' && s[end] == ')') || (end-start < 4)) {
        fprintf(stderr, "parse: unparsable: ");
        for (;start<=end;start++) {
            fprintf(stderr, "%c", s[start]);
        }
        fprintf(stderr, "\n");
        exit(-1);
    }

    // unitary operator
    if (s[start+1] == '~') {
        rval = parse(s, start+3, end-1);
        return new_node('u', '~', 0, NULL, rval);
    }

    // various binary operators
    if ( strchr("*/%+-&^|<>%", s[start+1]) != NULL) {
        cval = s[start+1];
        offset = 3;
        if (cval == '<' && s[start+2] == '<' && s[start+3] == ' ') {
            cval = 'l';
            offset = 4;
        } else if (cval == '>' && s[start+2] == '>' && s[start+3] == ' ') {
            cval = 'r';
            offset = 4;
        } else if (s[start+2] != ' ') {
            fprintf(stderr, "parse: invalid operator: %c\n", cval);
            exit(-1);
        }
        split = find_split(s, start+offset, end-1);
        lval = parse(s, start+offset, split-1);
        rval = parse(s, split+1, end-1);
        return new_node('b', cval, 0, lval, rval);
    }

    // should not get here 
    fprintf(stderr, "parse: feel through\n");
    exit(-1);
}

unsigned int execute(struct node *sexpr, unsigned int t) {
    switch (sexpr->type) {
    // atom
    case 'v':
        return t;
    case 'n':
        return sexpr->ival;
    // unary
    case 'u':
        if (sexpr->cval == '~') {
            return (~ execute(sexpr->rval, t));
        } else {
            fprintf(stderr, "unexpected unary oper: %c",  sexpr->cval);
        }
    // binary
    case 'b':
        switch (sexpr->cval) {
        case '+':
            return execute(sexpr->lval, t) + execute(sexpr->rval, t);
        case '-':
            return execute(sexpr->lval, t) - execute(sexpr->rval, t);
        case '*':
            return execute(sexpr->lval, t) * execute(sexpr->rval, t);
        case '/':
            return execute(sexpr->lval, t) / execute(sexpr->rval, t);
        case '^':
            return execute(sexpr->lval, t) ^ execute(sexpr->rval, t);
        case '&':
            return execute(sexpr->lval, t) & execute(sexpr->rval, t);
        case '|':
            return execute(sexpr->lval, t) | execute(sexpr->rval, t);
        case '%':
            return execute(sexpr->lval, t) % execute(sexpr->rval, t);
        case 'r':
            return execute(sexpr->lval, t) >> execute(sexpr->rval, t);
        case 'l':
            return execute(sexpr->lval, t) << execute(sexpr->rval, t);
        default:
            fprintf(stderr, "unexpected binary oper: %c", sexpr->cval);
            exit(-1);
        }
    default:
        fprintf(stderr, "execute: unknown type: %c", sexpr->type);
        exit(-1);
    }
}

void play(struct node *sexpr) {
    unsigned int t = 0;
    for (;;t++) {
        printf("%c", (char)(0x000000FF & execute(sexpr, t)));
    }
    return;
}

int find_split(char *s, int start, int end) {
    int depth = 0;
    int i;
    for (i=start; i <= end; i++) {
        if ((depth == 0) && (s[i] == ' '))
            return i;
        if (s[i] == '(')
            depth++;
        if (s[i] == ')')
            depth--; 
        if (depth < 0) {
            fprintf(stderr, "parse: unmatched ')'\n");
            exit(-1);
        }
    }
    if (depth > 0) {
        fprintf(stderr, "parse: unmatched '('\n");
        exit(-1);
    }
    fprintf(stderr, "parse: could not find split\n");
    exit(-1);
}

void print_sexpr(struct node *sexpr) {
    char oper = '_';
    char twice = 0;
    switch (sexpr->type) {
    // atoms
    case 'v':
        printf("t");
        break;
    case 'n':
        printf("%d", sexpr->ival);
        break;
    // unary operators
    case 'u':
        if (sexpr->cval == '~') {
            printf("(~ ");
            print_sexpr(sexpr->rval);
            printf(")");
        } else {
            fprintf(stderr, "unexpected unary: %c", sexpr->cval);
        }
    // binary operators
    case 'b':
        if ((sexpr->cval == 'l') || (sexpr->cval == 'r')) {
            twice = 1;
            oper = '<';
            if (sexpr->cval == 'r')
                oper = '>';
        } else {
            oper = sexpr->cval;
        }
        if (strchr("+-*/^|&<>", oper) != NULL) {
            printf("(%c ", oper);
            if (twice)
                printf("%c", oper);
            print_sexpr(sexpr->lval);
            printf(" ");
            print_sexpr(sexpr->rval);
            printf(")");
        } else {
            fprintf(stderr, "unexpected binary: %c", sexpr->cval);
            exit(-1);
        }

    }
}

struct node *new_node(char type, char cval, unsigned int ival, 
                      struct node *lval, struct node *rval) {
    struct node *n;
    newest_node++;
    if (newest_node >= NUMNODE) {
        fprintf(stderr, "node table overrun\n");
        exit(-1);
    }
    n = &node_table[newest_node];
    n->type = type;
    n->cval = cval;
    n->ival = ival;
    n->lval = lval;
    n->rval = rval;
    return n;
}

